Windows improvements

compnerd opened the pull request.

adierking commented on src/event/event_windows.c:
```diff
@@ -48,34 +60,193 @@ _dispatch_unote_unregister(dispatch_unote_t du DISPATCH_UNUSED,
 
 DISPATCH_NOINLINE
 void
 _dispatch_event_loop_drain(uint32_t flags DISPATCH_UNUSED)
```
I think we should remove the `DISPATCH_UNUSED` here because we check the `KEVENT_FLAG_IMMEDIATE` flag.

compnerd replied:
Well, that would just be asking that I learn to write code ... *sigh*.  Thanks

MadCoder commented on src/shims/lock.h:
```diff
@@ -285,6 +285,9 @@ _dispatch_thread_event_init(dispatch_thread_event_t dte)
 #else
 	_dispatch_sema4_init(&dte->dte_sema, _DSEMA4_POLICY_FIFO);
+#if defined(_WIN32)
+	_dispatch_sema4_create(&dte->dte_sema, _DSEMA4_POLICY_FIFO);
```
this seems wrong, the point of the sema4 is to be lazy and do the create() only if we're hitting the kernel
why can't windows do the same?

MadCoder replied:
Or alternatively, you should make `_dispatch_sema4_init` always create, which is how the posix semaphore works (`_dispatch_sema4_is_created` always says `true`).

compnerd replied:
Unfortunately I don't remember the test (it was in XCTest's test suite), but the queue wouldn't drain with the lazy semaphore.  I've adopted the posix semaphore policy instead as you suggested.  I think that it would be a good thing to do as a follow up (investigate and fix the lazy semaphores).

MadCoder commented on src/shims/lock.h:
```diff
@@ -285,6 +285,9 @@ _dispatch_thread_event_init(dispatch_thread_event_t dte)
 	dte->dte_value = 0;
 #else
 	_dispatch_sema4_init(&dte->dte_sema, _DSEMA4_POLICY_FIFO);
+#if defined(_WIN32)
```
you can remove this hunk now since _init() will create

compnerd replied:
Oops!

compnerd commented:
CC: @MadCoder @ktopley-apple - okay, with this patch set, I can pass the XCTest test suite (modulo some unrelated changes).

compnerd commented:
@MadCoder thanks for having a look!
@ktopley-apple - how should we handle this?  But, this definitely makes dispatch way better for Windows.  Most of this is relevant even after the 1121 merge.

adierking commented on src/shims/time.h:
```diff
@@ -108,12 +108,13 @@ _dispatch_get_nanoseconds(void)
 	li.LowPart = ft.dwLowDateTime;
 	li.HighPart = ft.dwHighDateTime;
-	return li.QuadPart * 100ull;
+	return li.QuadPart * 100ull + kNTToUNIXBiasAdjustment;
```
Shouldn't we be subtracting the bias here rather than adding?

compnerd replied:
Yeah, I thought I was subtracting.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

ktopley-apple commented:
> @MadCoder thanks for having a look!
> @ktopley-apple - how should we handle this? But, this definitely makes dispatch way better for Windows. Most of this is relevant even after the 1121 merge.

Once 1121 is completed, the other PRs can then be merged on top of it (assuming they still apply cleanly). I will be continuing work on the 1121 merge this week.

compnerd commented:
@ktopley-apple - okay, this definitely wont apply cleanly to 1121.  I'm definitely looking forward to the 1121 merge, but this is a fairly huge improvement to libdispatch on Windows, and is needed to get XCTest building and passing tests on Windows, so it would be really helpful to get this in (I'm okay with getting 1121 merged and then re-doing this patchset for 1121, as that has a bunch of improvements that I am really looking forward to).

compnerd commented:
@ktopley-apple - ugh, seems that the merge completely broke the Windows build for swift (as we cannot build dispatch any more).  I'm going to try to take a look at this today, but, unfortunately, initial testing did not go too smoothly.  Would you mind taking a look at the 1121 merge resolution (re-application of the changes that you did to the other targets).  Everything after @adierking's change is the stuff that is different, everything prior @MadCoder took a look at previously and just has been rebased (and a single patch was dropped).

ktopley-apple commented:
> @ktopley-apple - ugh, seems that the merge completely broke the Windows build for swift (as we cannot build dispatch any more). I'm going to try to take a look at this today, but, unfortunately, initial testing did not go too smoothly. Would you mind taking a look at the 1121 merge resolution (re-application of the changes that you did to the other targets). Everything after @adierking's change is the stuff that is different, everything prior @MadCoder took a look at previously and just has been rebased (and a single patch was dropped).

My impression was that the commit that you had on a separate branch, when applied on top of my -1121 branch, was sufficient to make the Windows port build. I didn't incorporate that commit because it had outstanding comments from @MadCoder that needed to be addressed. Is there something more that is broken?

compnerd commented:
@ktopley-apple - yeah, I had missed that message.  The set of changes from that are part of this branch, but I had in the mean time, also fixed timer queues on Windows to work properly (which they no longer do with 1121).  The set of changes were sufficient to get XCTest's test suite passing.  I think that I addressed most of the comments that @MadCoder had on that change in this version, and this does build, though, I need to get the timers working again.

ktopley-apple commented:
> @ktopley-apple - yeah, I had missed that message. The set of changes from that are part of this branch, but I had in the mean time, also fixed timer queues on Windows to work properly (which they no longer do with 1121). The set of changes were sufficient to get XCTest's test suite passing. I think that I addressed most of the comments that @MadCoder had on that change in this version, and this does build, though, I need to get the timers working again.

OK, good. Let me know as and when you need me to review patches and start merging changes.

compnerd commented:
@ktopley-apple - if you can look through the current patch set, I think that would be wonderful.  Even though the timers may not work with this version, this will repair the build of the swift compiler and running the swift standard library/compiler test suite.  If I can't get the timers working immediately, I would rather get that repaired and then follow up with the fixes for the timer queues.

ktopley-apple commented:
> @ktopley-apple - if you can look through the current patch set, I think that would be wonderful. Even though the timers may not work with this version, this will repair the build of the swift compiler and running the swift standard library/compiler test suite. If I can't get the timers working immediately, I would rather get that repaired and then follow up with the fixes for the timer queues.

For the avoidance of doubt, can you list the PRs that you want me to look at?

compnerd commented:
@ktopley-apple - sure, #453 I would say is the most important one, #454 needs some work, but that is the other one that is interesting.

ktopley-apple commented on private/private.h:
```diff
@@ -177,13 +177,13 @@ void _dispatch_prohibit_transition_to_multithreaded(bool prohibit);
 #endif
 
-#if DISPATCH_COCOA_COMPAT || defined(_WIN32)
+#if DISPATCH_COCOA_COMPAT
```
Great to see that you did this 👍

ktopley-apple commented on src/io.c:
```diff
@@ -1399,11 +1399,11 @@ _dispatch_fd_entry_create_with_fd(dispatch_fd_t fd, uintptr_t hash)
 			(void)dispatch_assume_zero(result);
 			_dispatch_stream_init(fd_entry,
-				_dispatch_get_root_queue(DISPATCH_QOS_DEFAULT, false));
+				_dispatch_get_default_queue(false));
```
@MadCoder commented on this in the review on the other branch. It's not clear what this change is for.

compnerd replied:
Ah, this is to mirror the changes that you did in the Linux side.

compnerd replied:
https://github.com/apple/swift-corelibs-libdispatch/commit/43a77d2bfae9e729e6de9b5b93ff985a2b5a2ec5#diff-592ee3fd43260e8d9bd24192211bce59L1558

ktopley-apple commented on src/io.c:
```diff
@@ -1399,11 +1399,11 @@ _dispatch_fd_entry_create_with_fd(dispatch_fd_t fd, uintptr_t hash)
-			dispatch_once_f(&_dispatch_io_devs_lockq_pred, NULL,
-					_dispatch_io_devs_lockq_init);
+			dispatch_once_f(&_dispatch_io_init_pred, NULL,
+					_dispatch_io_queues_init);
```
Similarly, can you explain this change?

compnerd replied:
This also mirrors the Linux changes.

compnerd replied:
https://github.com/apple/swift-corelibs-libdispatch/commit/43a77d2bfae9e729e6de9b5b93ff985a2b5a2ec5#diff-592ee3fd43260e8d9bd24192211bce59L489

ktopley-apple commented:
> @ktopley-apple - sure, #453 I would say is the most important one, #454 needs some work, but that is the other one that is interesting.

OK, I had some comments on #453. Let me know when #454 works and is ready for review.

compnerd commented:
@ktopley-apple - if there are reasons to do the changes differently, please let me know, and Ill update the diffs.

ktopley-apple commented:
> @ktopley-apple - if there are reasons to do the changes differently, please let me know, and Ill update the diffs.

All looks good, thanks.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@ktopley-apple - okay, so, I've pulled in the changes from @adierking into this set to run the tests as well.  There was a dropped change that caused the failure, but this now builds and passes the timer queue tests as well!  With this set of changes, libdispatch 1121 should be as good as 931 on Windows (well, better than what was actually in tree).  This should be ready to merge now :-)

ktopley-apple commented:
> @ktopley-apple - okay, so, I've pulled in the changes from @adierking into this set to run the tests as well. There was a dropped change that caused the failure, but this now builds and passes the timer queue tests as well! With this set of changes, libdispatch 1121 should be as good as 931 on Windows (well, better than what was actually in tree). This should be ready to merge now :-)

Great! I'll merge it once the build is clean.

compnerd commented:
@ktopley-apple - awesome, seems that this is through the build bot, and should greatly improve libdispatch on Windows!  Thanks for the help with that
